File: 0015-3sum/0015-3sum.py
class Solution(object):
    def threeSum(self, nums):
        nums.sort()
        n = len(nums)
        ans = []
        for i in range(n):
            if i!=0 and nums[i] == nums[i-1]:
                continue
            j = i+1
            k = n-1
            while j < k :
                sum = nums[i]+nums[j]+nums[k]
                if sum < 0 :
                    j = j+1
                elif sum > 0 :
                    k = k-1
                else : 
                    temp = [nums[i],nums[j],nums[k]]
                    ans.append(temp)
                    j = j+1
                    k = k-1
                    while j < k and nums[j] == nums[j-1] :
                        j = j+1
                    while j < k and nums[k] == nums[k+1] :
                        k = k-1
        return ans

obj = Solution()
nums = [-1,0,1,2,-1,-4]
obj.threeSum(nums)

File: 0015-3sum/0015-3sum.java
class Solution {
    public void twoSum(List<List<Integer>> ans, int index, int target, int nums[] ){
        int i = index;
        int j = nums.length - 1;
        while(i < j){
            int b = nums[i];
            int c = nums[j];
            if(b + c == target){
                List<Integer> list = new ArrayList<>();
                list.add(-target);
                list.add(b);
                list.add(c);
                ans.add(list);
                i ++;
                while(i < j && nums[i] == nums[i - 1])
                    i ++;
                j --;
                while(i < j && nums[j] == nums[j + 1])
                    j --;
            }
            else if(b + c > target)
                j --;
            else
                i ++;
        }
    }
    public List<List<Integer>> threeSum(int[] nums) {
       Arrays.sort(nums);
       List<List<Integer>> ans = new ArrayList<>();
       for(int i = 0; i < nums.length - 2; i ++){
            if(i == 0 || nums[i] != nums[i - 1]){
                int a = nums[i];
                twoSum(ans, i + 1, -a, nums);
            }
       }
       return ans;
    }
}